### PR TITLE
system/update : Add support of binary update

### PR DESCRIPTION
**Pull request by YiHyunJin**
Support api for binary update.

load_new_binary :
This function is a function that loads a new binary.

get_binary_info :
This function get with binary information.

get_binary_info_all :
This function get with all binary information.

clean_binary_info_list :
This function clears binary information list.

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [6c08d6578aa2754e277c8ee872526106859056a2] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [6c08d6578aa2754e277c8ee872526106859056a2] - Code Rule Check Result:
apps/system/update/binary_update.h:39: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
os/kernel/binary_manager/binary_manager_getinfo.c:118: ERROR: [SPC_M_OPR] space required after that ',' (ctx:WxV)
os/kernel/binary_manager/binary_manager_getinfo.c:154: ERROR: [SPC_M_SEP] space prohibited after that open parenthesis '('

File: apps/system/update/Make.defs [hunk 1]
```diff
@@ -0,0 +1,22 @@
+############################################################################
+#
+# Copyright (c) 2017 Samsung Electronics Co., Ltd All Rights Reserved
```
Comment (NilabhShant) [inline apps/system/update/Make.defs]: 2017 --> 2019

File: apps/system/update/binary_update.c [hunk 2]
```diff
@@ -0,0 +1,237 @@
+/****************************************************************************
+ *
+ * Copyright 2017 Samsung Electronics All Rights Reserved.
```
Comment (NilabhShant) [inline apps/system/update/binary_update.c]: same as above.

File: apps/system/update/binary_update.c [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+	binmgr_mq = mq_open(BINMGR_REQUEST_MQ, O_WRONLY, 0666, 0);
+	if (binmgr_mq == (mqd_t)ERROR) {
+		dbg("mq_open failed!\n");
+		return ERROR;
+	}
+
+	ret = mq_send(binmgr_mq, (const char *)request_msg, sizeof(binmgr_request_t), BINMGR_NORMAL_PRIO);
+	if (ret == ERROR) {
+		dbg("mq_send failed!\n");
+		return ERROR;
```
Comment (yashwanth686007) [inline apps/system/update/binary_update.c]: Please close the queue before returning error.

File: apps/system/update/binary_update.c [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+
+	while (response_msg.cmd == REQUESET_CONTINUE) {
+		ret = receive_response(private_mq, &response_msg);
+		if (ret == ERROR) {
+			binary_list_clean(binary_info_list);
+			mq_close(private_mq);
+			mq_unlink(request_msg.q_name);
+			return ERROR;
+		}
+		next_list->next = (binary_info_list_t *)malloc(sizeof(binary_info_list_t));
```
Comment (yashwanth686007) [inline apps/system/update/binary_update.c]: Please check whether malloc is successful or not.

**Comment by seinfra** [comment]
Target : [8f3502c1122b23648aa02f2790f5058509f33de3] - Code Rule Check Result:
apps/system/update/binary_update.h:39: ERROR: [IDT_M_TAB] please, no spaces at the start of a line

**Comment by seinfra** [comment]
Target : [8f3502c1122b23648aa02f2790f5058509f33de3] - Code Rule Check (C++) OK.

File: apps/system/update/binary_update.c [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+
+	mq_close(private_mq);
+	mq_unlink(request_msg.q_name);
+	return OK;
+}
+
+void binary_list_clean(binary_info_list_t *binary_info_list)
+{
+	if (binary_info_list->next != NULL) {
+		binary_list_clean(binary_info_list->next);
```
Comment (yashwanth686007) [inline apps/system/update/binary_update.c]: Instead of calling recursively, the list can be traversed linearly and cleaned up.

File: apps/system/update/binary_update.h [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+struct binary_info_s {
+	int part_size;
+	char name[BIN_NAME_MAX];
+	char version[BIN_VERSION_MAX];
+	char dev_path[BINMGR_DEVNAME_LEN];
+};
+typedef struct binary_info_s binary_info_t;
+
+struct binary_info_list_s {
+	binary_info_t binary_info;
```
Comment (yashwanth686007) [inline apps/system/update/binary_update.h]: Please check for alignment.

**Comment by seinfra** [comment]
Target : [ea591dc0829f47dc1c7d1d96409f053bb3cc837c] - Code Rule Check Result:
apps/system/update/binary_update.h:39: ERROR: [IDT_M_TAB] please, no spaces at the start of a line

File: os/kernel/binary_manager/binary_manager_getinfo.c [hunk 7]
```diff
@@ -79,13 +89,81 @@ int binary_manager_get_index_with_name(char *bin_name)
...
+
+	struct mq_attr attr;
+	attr.mq_maxmsg = BINMGR_MAX_MSG;
+	attr.mq_msgsize = sizeof(binmgr_response_t);
+	attr.mq_flags = 0;
+
+	private_mq = mq_open(q_name, O_WRONLY | O_CREAT, 0666, &attr);
+
+	if (private_mq == (mqd_t)ERROR) {
+		//tmdbg("mq_open failed!\n");
```
Comment (yashwanth686007) [inline os/kernel/binary_manager/binary_manager_getinfo.c]: It's better to add a debug comment.

**Comment by seinfra** [comment]
Target : [ea591dc0829f47dc1c7d1d96409f053bb3cc837c] - Code Rule Check (C++) OK.

File: os/kernel/binary_manager/binary_manager_getinfo.c [hunk 8]
```diff
@@ -79,13 +89,81 @@ int binary_manager_get_index_with_name(char *bin_name)
...
+			response_msg.cmd = REQUESET_DONE;
+		}
+
+		strncpy(response_msg.name, BIN_NAME(index), BIN_NAME_MAX);
+		response_msg.part_size = BIN_PARTSIZE(index);
+		strncpy(response_msg.version, BIN_VER(index), BIN_VER_MAX);
+		bmvdbg("bin_idx - %d, use - %d >> %d", index, BIN_USEIDX(index), BIN_PARTNUM(index, BIN_USEIDX(index)));
+		bmvdbg("notuse - %d >> %d\n", (BIN_USEIDX(index) ^ 1), BIN_PARTNUM(index, (BIN_USEIDX(index) ^ 1)));
+		snprintf(response_msg.dev_path, BINMGR_DEVNAME_LEN, BINMGR_DEVNAME_FMT, BIN_PARTNUM(index, (BIN_USEIDX(index) ^ 1)));
+		ret = mq_send(private_mq, (const char *)&response_msg, sizeof(response_msg), 0);
```
Comment (yashwanth686007) [inline os/kernel/binary_manager/binary_manager_getinfo.c]: Please check for return value.

**Comment by seinfra** [comment]
Target : [4beb3fd73cda4ff0e5f13b359f4fce1a60bc8338] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [4beb3fd73cda4ff0e5f13b359f4fce1a60bc8338] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [65f78c345ebe9c46506558269c4c41aa0ccea427] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [65f78c345ebe9c46506558269c4c41aa0ccea427] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [435cffd09265e833eb358a945f35dd955421d219] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [435cffd09265e833eb358a945f35dd955421d219] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [43b3e4f8a706a477326d0980cd1611308d812e1b] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [43b3e4f8a706a477326d0980cd1611308d812e1b] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [22f83c2aefc073c134182a5abb2a1844a3babd69] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [22f83c2aefc073c134182a5abb2a1844a3babd69] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [04ea0d08f90ca5dfad60aaac8617b03405bacdf7] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [04ea0d08f90ca5dfad60aaac8617b03405bacdf7] - Code Rule Check OK.

File: apps/system/update/binary_update.c [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+/***************************************************************************
+ * Included Files
+ ***************************************************************************/
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <errno.h>
+#include <sched.h>
+#include <time.h>
```
Comment (thapav) [inline apps/system/update/binary_update.c]: please remove unwanted includes

File: apps/system/update/binary_update.c [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+#include <fcntl.h>
+#include <string.h>
+#include <pthread.h>
+#include <sys/time.h>
+#include <mqueue.h>
+#include <tinyara/binary_manager.h>
+
+#include "binary_update.h"
+
+static int send_requset(binmgr_request_t *request_msg)
```
Comment (thapav) [inline apps/system/update/binary_update.c]: send_request*

File: apps/system/update/binary_update.c [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+		return ERROR;
+	}
+
+	memset((void *)&request_msg, 0, sizeof(binmgr_request_t));
+
+	request_msg.cmd = BINMGR_RELOAD;
+	request_msg.pid = getpid();
+	snprintf(request_msg.b_name, BIN_NAME_MAX, "%s", binary_name);
+
+	ret = send_requset(&request_msg);
```
Comment (thapav) [inline apps/system/update/binary_update.c]: correct name of API in all used places#

File: apps/system/update/binary_update.c [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+
+#include "binary_update.h"
+
+static int send_requset(binmgr_request_t *request_msg)
+{
+	int ret;
+	mqd_t binmgr_mq;
+	binmgr_mq = mq_open(BINMGR_REQUEST_MQ, O_WRONLY, 0666, 0);
+	if (binmgr_mq == (mqd_t)ERROR) {
+		mq_close(binmgr_mq);
```
Comment (sunghan-chang) [inline apps/system/update/binary_update.c]: unnecessary

File: apps/system/update/binary_update.c [hunk 13]
```diff
@@ -0,0 +1,248 @@
...
+		dbg("mq_open failed!\n");
```
Comment (sunghan-chang) [inline apps/system/update/binary_update.c]: You should not use `dbg`. You should make this module specific debug messages or use printf.

File: apps/system/update/binary_update.c [hunk 14]
```diff
@@ -0,0 +1,248 @@
...
+		return ERROR;
+	}
+
+	ret = mq_send(binmgr_mq, (const char *)request_msg, sizeof(binmgr_request_t), BINMGR_NORMAL_PRIO);
+	if (ret == ERROR) {
```
Comment (sunghan-chang) [inline apps/system/update/binary_update.c]: ret is required? This ret is used only here.
How about below without ret?
```
if (mq_send(xxx) == ERROR) {
```

File: apps/system/update/binary_update.c [hunk 15]
```diff
@@ -0,0 +1,248 @@
...
+	return;
+}
+
+int load_new_binary(char *binary_name)
+{
+	int ret;
+	binmgr_request_t request_msg;
+
+	if (binary_name == NULL) {
+		dbg("load_new_binary failed binary nams is NULL\n");
```
Comment (sunghan-chang) [inline apps/system/update/binary_update.c]: name*

File: apps/system/update/binary_update.h [hunk 16]
```diff
@@ -0,0 +1,115 @@
+/****************************************************************************
```
Comment (sunghan-chang) [inline apps/system/update/binary_update.h]: You should put this under apps/include/system folder to allow to call belows.

File: apps/system/update/binary_update.c [hunk 17]
```diff
@@ -0,0 +1,248 @@
...
+		return ERROR;
+	}
+
+	memset((void *)&request_msg, 0, sizeof(binmgr_request_t));
```
Comment (sunghan-chang) [inline apps/system/update/binary_update.c]: Why is this required? It looks unnecessary.

File: apps/system/update/binary_update.c [hunk 18]
```diff
@@ -0,0 +1,248 @@
...
+	binmgr_response_t response_msg;
+	mqd_t private_mq;
+	struct mq_attr attr;
+
+	if (binary_name == NULL) {
+		dbg("get_binary_info failed. binary nams is NULL\n");
+		return ERROR;
+	}
+
+	memset((void *)&request_msg, 0, sizeof(binmgr_request_t));
```
Comment (sunghan-chang) [inline apps/system/update/binary_update.c]: Why is this required? To add null at end of q_name and b_name?

File: apps/system/update/binary_update.c [hunk 19]
```diff
@@ -0,0 +1,248 @@
...
+		dbg("mq_send failed!\n");
+		return ERROR;
+	}
+
+	attr.mq_maxmsg = BINMGR_MAX_MSG;
+	attr.mq_msgsize = sizeof(binmgr_response_t);
+	attr.mq_flags = 0;
+
+	private_mq = mq_open(request_msg.q_name, O_RDONLY | O_CREAT, 0666, &attr);
+
```
Comment (sunghan-chang) [inline apps/system/update/binary_update.c]: Why do you make an empty line only here?

File: os/include/tinyara/binary_manager.h [hunk 20]
```diff
@@ -67,22 +73,33 @@ enum binmgr_request_msg_type {
 #endif
 };
 
+/* Response essage type for binary manager */
```
Comment (jsdosa) [inline os/include/tinyara/binary_manager.h]: could you change essage to message

File: os/include/tinyara/binary_manager.h [hunk 21]
```diff
@@ -67,22 +73,33 @@ enum binmgr_request_msg_type {
...
+enum binmgr_response_msg_type {
+	REQUESET_DONE,
```
Comment (jsdosa) [inline os/include/tinyara/binary_manager.h]: could you change REQUESET to REQUEST ?

File: apps/system/update/binary_update.c (hunk 3 above)
Comment (YiHyunJin) [reply apps/system/update/binary_update.c]: Done.

**Comment by seinfra** [comment]
Target : [e0e8dec8163fbe7863fd568a23b3314d615f9983] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [e0e8dec8163fbe7863fd568a23b3314d615f9983] - Code Rule Check OK.

File: apps/system/update/binary_update.c [hunk 22]
```diff
@@ -0,0 +1,240 @@
...
+	nbytes = mq_receive(private_mq, (char *)response_msg, sizeof(binmgr_response_t), 0);
+	if (nbytes <= 0) {
+		bmdbg("receive ERROR %d, errno %d, retry!\n", nbytes, errno);
+		return ERROR;
+	}
+
+	return OK;
+}
+
+static void binary_info_copy(binary_info_t *binary_info, binmgr_response_t *response_msg)
```
Comment (jeongchanKim) [inline apps/system/update/binary_update.c]: `copy_binary_info` is better.

File: apps/system/update/binary_update.c [hunk 23]
```diff
@@ -0,0 +1,240 @@
...
+		return ERROR;
+	}
+
+	attr.mq_maxmsg = BINMGR_MAX_MSG;
+	attr.mq_msgsize = sizeof(binmgr_response_t);
+	attr.mq_flags = 0;
+
+	private_mq = mq_open(request_msg.q_name, O_RDONLY | O_CREAT, 0666, &attr);
+	if (private_mq == (mqd_t)ERROR) {
+		mq_close(private_mq);
```
Comment (jeongchanKim) [inline apps/system/update/binary_update.c]: mq_close and mq_unlink is not needed.

File: apps/system/update/binary_update.c [hunk 24]
```diff
@@ -0,0 +1,240 @@
...
+	struct mq_attr attr;
+
+	memset((void *)&request_msg, 0, sizeof(binmgr_request_t));
+
+	request_msg.cmd = BINMGR_GET_INFO_ALL;
+	request_msg.pid = getpid();
+
+	snprintf(request_msg.q_name, BIN_NAME_MAX, "%s%d", BINMGR_RESPONSE_MQ, request_msg.pid);
+
+	send_request(&request_msg);
```
Comment (jeongchanKim) [inline apps/system/update/binary_update.c]: why didn't you check the return value?

File: apps/system/update/binary_update.h [hunk 25]
```diff
@@ -0,0 +1,115 @@
...
+typedef struct binary_info_s binary_info_t;
+
+struct binary_info_list_s {
+	binary_info_t binary_info;
+	struct binary_info_list_s* next;
+};
+typedef struct binary_info_list_s binary_info_list_t;
+
+
+/****************************************************************************
```
Comment (jeongchanKim) [inline apps/system/update/binary_update.h]: I think this explanation should be changed to doxygen format.

File: apps/system/update/binary_update.c (hunk 4 above)
Comment (YiHyunJin) [reply apps/system/update/binary_update.c]: Done.

**Comment by seinfra** [comment]
Target : [265ea301264c2ec324810d51830771c5880bc4dd] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [265ea301264c2ec324810d51830771c5880bc4dd] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [fa84d6ddcdb6a198e0fca1d546870e3198d281c4] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [fa84d6ddcdb6a198e0fca1d546870e3198d281c4] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [6d83cc580c728de301c944ec898193c5a653eca1] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [6d83cc580c728de301c944ec898193c5a653eca1] - Code Rule Check OK.
